Improve documentation around func and method (#19207)
nim-lang/Nim: doc/tut1.rst

To define new commands like `echo <system.html#echo,varargs[typed,]>`_
and `readLine <io.html#readLine,File>`_ in the examples, the concept of a
*procedure* is needed. You might be used to them being called *methods* or
*functions* in other languages, but Nim
`differentiates these concepts <tut1.html#procedures-funcs-and-methods>`_. In

> Nim new procedures are defined with the `proc` keyword:

Reviewer: ```suggestion
Nim, new procedures are defined with the `proc` keyword:
```

whose value is then returned implicitly.

Funcs and methods
-----------------

> As mentioned in the introduction Nim differentiates between procedures,

Reviewer: ```suggestion
As mentioned in the introduction, Nim differentiates between procedures,
```

functions, and methods, defined by the `proc`, `func`, and `method` keywords

> respectively. In some ways Nim are a bit more pedantic in its definitions than

Reviewer 2: ```suggestion
respectively. In some ways, Nim is a bit more pedantic in its definitions than
```

> othe languages.

Reviewer 2: ```suggestion
other languages.
```

Functions are closer to the concept of a pure mathematical
function, which might be familiar to you if you've ever done functional

> programming. Essentially they can't access global state (except `const`), can't

Reviewer: ```suggestion
programming. Essentially they can't access global state (except `const`) and can't
```

> programming. Essentially they can't access global state (except `const`) and can't
> produce side-effects (they are tagged with `{.noSideEffects.}`). They can still

Reviewer 3: ```suggestion
programming. Essentially they are procedures with additional limitations set on them:
they can't access global state (except `const`) and can't produce side-effects. `func`
keyword is basically an alias for the `proc` tagged with `{.noSideEffects.}` pragma. Functions can still
```

change their mutable arguments, which are those marked as `var` and any `ref`
objects.

> Unlike procedures methods are dynamically dispatched. This sounds a bit

Reviewer: ```suggestion
Unlike procedures, methods are dynamically dispatched. This sounds a bit
```

> complicated but is a concept closely related to inheritance and object oriented

Reviewer: ```suggestion
complicated, but is a concept closely related to inheritance and object oriented
```

> complicated, but is a concept closely related to inheritance and object oriented

Reviewer 3: ```suggestion
complicated, but it is a concept closely related to inheritance and object oriented
```

programming. If you overload a procedure (two procedures with the same name but

> different types are said to be overloaded) the procedure to use is determined

Reviewer: ```suggestion
different types are said to be overloaded), the procedure to use is determined
```

> different types are said to be overloaded), the procedure to use is determined

Reviewer 3: ```suggestion
of different types or with different sets of arguments are said to be overloaded), the procedure to use is determined
```
Being pedantic, a proc with different args is of a different type by definition, but this wording captures both the args and the result type unambiguously.

> at compile-time. Methods on the other hand depend on objects that inherit from

Reviewer 3: ```suggestion
at compile-time. Methods, on the other hand, depend on objects that inherit from
```

> the `RootObj`. This is something that will be covered in much greater depth in

Reviewer 3: ```suggestion
the `RootObj`. This is something that is covered in much greater depth in
```
This is not a blogpost, tut2 is already written. :)

Author: I was more thinking of it as being read linearly, so while you're reading tut1 then tut2 is still in the future. But I don't mind either way

whose value is then returned implicitly.

Funcs and methods
-----------------

As mentioned in the introduction, Nim differentiates between procedures,
functions, and methods, defined by the `proc`, `func`, and `method` keywords
respectively. In some ways, Nim is a bit more pedantic in its definitions than
other languages.

> Functions are closer to the concept of a pure mathematical
> function, which might be familiar to you if you've ever done functional
> programming. Essentially they are procedures with additional limitations set on
> them: they can't access global state (except `const`) and can't produce
> side-effects. The `func` keyword is basically an alias for `proc` tagged
> with `{.noSideEffects.}`. Functions can still change their mutable arguments
> however, which are those marked as `var`, along with any `ref` objects.

Reviewer 3: ```suggestion
Functions are essentially procedures with additional limitations set on them:
they can't access global state (except `const`) and can't produce side-effects.
This puts Nim functions closer to the concept of a mathematical function, which
might be familiar to you if you've ever done functional programming. The main
benefit of the imposed limits is that it's much easier to reason about the code
with minimized interdependence of its various parts, as data flows directly
through a function from its arguments to its output. The other notable advantage
is that the self-contained nature of a function makes it inherently more
portable and easier to test.

Even though these benefits apply to Nim functions, it's important to keep in
mind that a `func` can still change its mutable arguments: those marked as `var`
along with `ref` objects (which possess interior mutability). This still allows
functions to have additional outputs beside the return value, making them not as
"pure" as their mathematical counterparts.

Technically, the `func` keyword is a shorthand alias for `proc` tagged with
`{.noSideEffects.}`.
```

Reviewer: One small typo: it's "possess", not "posses".

> side-effects. The `func` keyword is basically an alias for the `proc` tagged
> with `{.noSideEffects.}`). Functions can still change their mutable arguments
> however, which are those marked as `var`, along with any `ref` objects.

Reviewer: ```suggestion
side-effects. The `func` keyword is basically an alias for `proc` tagged
with `{.noSideEffects.}`. Functions can still change their mutable arguments
however, which are those marked as `var`, along with any `ref` objects.
```

Reviewer 3: That's just getting better and better!
